### ISSUE TITLE
ONNX to RKNN conversion crashes/fails with UINT8/INT8 input type

### ISSUE THREAD

**Issue by topiolli** [airockchip/rknn-toolkit2#526]
**Bug description**

I'm unable to create an RKNN model that would accept 8-bit input data. It seems to me that this should be possible as the rknpu examples come with models that accept 8-bit input.¹

I created a minimal repro model that just casts input to float. If I run the code below, I get the following behavior:

1. input type = `float`: Works
2. input type = `int8`: Segmentation fault
3. input type = `uint8`: `E RKNN: [07:41:15.760] Not Support Dtype: 2. please check precision!`

In `rknn_api.h`, `RKNN_TENSOR_INT8` equals 2, but the input type is unsigned. Either the error message is misleading or the enum value is wrong.

How can I create a model with 8-bit input? 

**Environment**

- `uname -a`: `Linux ubuntu 5.10.0-1012-rockchip #12-Ubuntu SMP Wed Aug 14 22:22:22 UTC 2024 aarch64 aarch64 aarch64 GNU/Linux`
- rknn_api/rknnrt version: 2.3.2 (429f97ae6b@2025-04-09T09:09:27)
- driver version: 0.9.6

**Footnotes**

1. I'm a bit baffled here. The mobilenet_v1.rknn model input type is `INT8`, not `UINT8`, but the example code just copies raw (unsigned) image data buffer to the input tensor. Are the values of `RKNN_TENSOR_INT8` and `RKNN_TENSOR_UINT8`  correct in the `rknn_tensor_type` enum? If they were swapped, the observed behavior would make more sense to me.

```py
# Minimal repro
import torch
from rknn.api import RKNN

class Model(torch.nn.Module):
    def forward(self, x):
        return x.float()

model = Model()
model.eval()

dummy_input = torch.randint(0, 256, (1, 3, 244, 244), dtype=torch.uint8)
#dummy_input = torch.randint(-128, 127, (1, 3, 244, 244), dtype=torch.int8)
#dummy_input = torch.randint(0, 256, (1, 3, 244, 244), dtype=torch.float32)

torch.onnx.export(
    model=model,
    args=dummy_input,
    f="/tmp/dummy.onnx",
    opset_version=15,
    input_names=['input'],
    output_names=['output'],
)

rknn = RKNN()
rknn.config(target_platform='rk3588')

def convert_onnx(onnx_file):
    ret = rknn.load_onnx(model=onnx_file)
    if ret != 0:
        print('Loading ONNX model failed!')
        return
    print('Building RKNN model...')
    ret = rknn.build(do_quantization=False)
    if ret != 0:
        print('Building RKNN model failed!')
        return

    rknn_model_path = onnx_file.replace('.onnx', '.rknn')
    print(f'Exporting RKNN model to {rknn_model_path}...')
    ret = rknn.export_rknn(rknn_model_path)
    if ret != 0:
        print('Exporting RKNN model failed!')

convert_onnx("/tmp/dummy.onnx")
rknn.release()
```

**Comment by topiolli** [comment]
I found a partial work-around that got me a bit further:

1. Use int8 as the input type when saving ONNX. This is wrong, but apparently cancels an input type detection bug in rknn toolkit.
2. Specify inputs explicitly and don't let the converter read them from ONNX. This got me past the segmentation fault.

```py
ret = rknn.load_onnx(model=onnx_file, inputs=['pixel_values'], input_size_list=[[1, 3, 244, 244]])
```

Ironically, this causes a warning:

`W load_onnx: If you don't need to crop the model, don't set 'inputs'/'input_size_list'/'outputs'!`

The minimal repro now passes without crashes, but I'm unsure about the result. Casting from int8 to float is different from casting to uint8 to float. I haven't yet tested if it actually works.

Edit: I tested and it did not work. Or rather did, but didn't do what I meant. The input seems to be actually int8 at run time, which isn't going to work with image input. If I feed the dummy network a tensor filled with `0xff` as int8, I get out -1 as float16. If I however change the input tensor type to uint8, I get 127 as float16. It seems that the input is clamped. This was a dead end. **How do I create a model with uint8 input type?**

Worse, I'm still unable to verify the result with my actual ConvNeXt model due to another bug:

`E RKNN: [09:02:14.219] Unkown op target: 0`

I'm filing another report for this, but what does it mean?

**Comment by topiolli** [comment]
According to this comment in the example source code, it should be possible to use uint8 directly:

```c++
// default input type is int8 (normalize and quantize need compute in outside)
// if set uint8, will fuse normalize and quantize to npu
```

https://github.com/airockchip/rknn-toolkit2/blob/master/rknpu2/examples/RV1106_RV1103/rknn_mobilenet_demo/src/main_nhwc.cc#L333-L334

That example sets the input tensor's type to uint8:

https://github.com/airockchip/rknn-toolkit2/blob/master/rknpu2/examples/RV1106_RV1103/rknn_mobilenet_demo/src/main_nhwc.cc#L320

The same pattern appears in other demos. However, the actual input type of the mobilenet_v1 RKNN model is *signed* int8. In my simple example feeding uint8 data to int8 input causes causes values over 127 to be lost due to clamping. What's going on here?

**Comment by happyme531** [comment]
Why you want to use INT8 input for your models? The operations like Conv will be executed in fp16 anyway so it's a waste of compute to do a conversion. 

The toolchain will convert the inputs to int8 if you choose to do quantization inside the toolchain. 

No matter what dtype the final rknn model have, you can feed fp32,fp16,int8 or any dtype in runtime, the NPU runtime will do an automatic conversion for you.

**Comment by topiolli** [comment]
Thanks for the reply, @happyme531!

I want to use UINT8 (not INT8) input because that is the native data type of images.

- If my model takes FLOAT32 input before rknn conversion with quantization disabled, the converted model will take FLOAT16 input, which requires conversion outside of the model and effectively prevents zero-copy.
- If I *do* quantize the model, the input will be INT8, which is again incompatible as detailed above. The automatic conversion is a problem because UINT8 input fed to an INT8 model seems to cause values above 127 to be clamped. At least that happens with my little test program above.

I tried to solve the problem by adding an explicit cast (UINT8 -> FLOAT32) to the front of the model. UINT8 input isn't however supported by the conversion tool.  What puzzles me is that the example programs seem to be doing the wrong thing: they happily feed raw UINT8 tensors to a model whose input is INT8.

What do you mean by saying that Conv will be executed in fp16 anyway? Do you mean that it is the default or that it *can* be done using hybrid quantization?

**Comment by topiolli** [comment]
After a few more hours of reverse engineering I got a bit further. If I modify the native input attribute types reported by the model – by changing `type` from `RKNN_TENSOR_FLOAT16` to `RKNN_TENSOR_UINT8` and fixing the buffer size (`size` and `size_with_stride`) accordingly – the model runs at least in a simple test case. But I *have to* specify mean and std during the export phase so that the input won't get clamped.

Please correct me if I'm wrong, but it seems that the normalization step in front of the actual model accepts tensor types that differer from what the model reports. The normalization gets applied even in zero-copy mode unless I explicitly set `pass_through` to one. ~It is a bit scary because it isn't clear from the documentation that modifying the input tensor types reported by the model would be safe.~ Not correct. See below.

**Comment by topiolli** [comment]
From the user guide, chapter **C API Zero-Copy Usage** on page 67:

> Special attention: If the data type of the query is uint8, but the user wants to pass in float32 type, the size of the rknn_tensor_attr structure must be modified to four times the original size, and the data type must be modified to RKNN_TENSOR_FLOAT32. After modification in this way, the hardware efficiency is not optimal, and the CPU will be called internally in the interface to perform data type conversion.

Three observations:

1. It seems that it should be possible to create a model with `uint8` input.
2. Feeding in a tensor that has a different data type than the reported one seems to cause a conversion on the CPU, which isn't what I'd call "zero-copy".
3. My previous comment about modifying the reported input types actually gets answered by this: it is safe.

The question still remains: how do I create a model that natively accepts `uint8` input?

**Comment by Lhy-opt** [comment]
Same, I casts model input to float thus avoiding this problem. The maintainers of this framework seem to care little about GitHub, accumulated too many issues. I've given up on this framework, same advice to you
